girder/models/collection.py

```diff
@@ -191,18 +192,74 @@ def fileList(self, doc, user=None, path='', includeMetadata=False,
         is counted as well.
 
         :param doc: The collection.
+        :param includeItems: Whether items should be included in the count.
+        :type includeItems: bool
+        :param user: If filtering by permission, the user to filter against.
+        :param level: If filtering by permission, the required permission level.
+        :type level: AccessLevel
         """
         count = 1
         folders = self.model('folder').find({
             'parentId': doc['_id'],
             'parentCollection': 'collection'
         })
-        count += sum(self.model('folder').subtreeCount(folder)
-                     for folder in folders)
+
+        if level is not None:
+            folders = self.filterResultsByPermission(
+                cursor=folders, user=user, level=level, limit=None)
+        count += sum(self.model('folder').subtreeCount(
+            folder, includeItems=includeItems, user=user, level=level)
+            for folder in folders)
         return count
+
+    def setAccessList(self, doc, access, save=False, recurse=False, user=None,
+                      progress=noProgress, setPublic=None):
+        """
+        Overrides AccessControlledModel.setAccessList to add a recursive
+        option. When `recurse=True`, this will set the access list on all
+        subfolders to which the given user has ADMIN access level. Any
+        subfolders that the given user does not have ADMIN access on will be
+        skipped.
+
+        :param doc: The collection to set access settings on.
+        :type doc: collection
+        :param access: The access control list.
+        :type access: dict
+        :param save: Whether the changes should be saved to the database.
+        :type save: bool
+        :param recurse: Whether this access list should be propagated to all
+            folders underneath this collection.
+        :type recurse: bool
+        :param user: The current user (for recursive mode filtering).
+        :param progress: Progress context to update.
+        :type progress: :py:class:`girder.utility.progress.ProgressContext`
+        :param setPublic: Pass this if you wish to set the public flag on the
+            resources being updated.
+        :type setPublic: bool or None
+        """
+        progress.update(increment=1, message='Updating ' + doc['name'])
+        if setPublic is not None:
+            self.setPublic(doc, setPublic, save=False)
+        doc = AccessControlledModel.setAccessList(self, doc, access, save=save)
+
+        if recurse:
+            cursor = self.model('folder').find({
+                'parentId': doc['_id'],
+                'parentCollection': 'collection'
+            })
+
+            folders = self.filterResultsByPermission(
+                cursor=cursor, user=user, level=AccessType.ADMIN, limit=None)
+
+            for folder in folders:
+                self.model('folder').setAccessList(
+                    folder, access, save=True, recurse=True, user=user,
```

Comment:
At this point you are saving the subtree objects regardless of the value of the method param `save`.  

This is probably better than setting `save=save` as it prevents some accidental behavior of not saving the subtree objects externally, at the possible cost of performance.  Most likely this would be called with `save=True` anyway (as it is in the only places it is currently called).

:heavy_check_mark: